<a href="https://colab.research.google.com/github/kyoungyi/Small-Projects/blob/master/all_county_corn_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
! pip install cftime
import xarray as xr
import numpy as np
import pandas as pd
import scipy.io
from itertools import zip_longest


     |████████████████████████████████| 296kB 2.9MB/s 


# 1. Crop Data
</li> source: USDA NASS () </li>
<li> temporal period: 1919 ~ 2019 (lots of data are missing depending on crop) </li>
<li> spatial scale: county level (data availability is different depending on crop) </li>
<li> note: missing data as zero </li>



In [2]:
data = scipy.io.loadmat('/content/drive/My Drive/all_crops_LLSupdated07282020_statesAndCounties.mat', squeeze_me=True)
county = data['st']['c']
#a.dtype.names
crop_names = county[0].dtype.names

In [11]:
#time = xr.cftime_range(start="1910-01-01", periods=110, freq="1Y", calendar='noleap')
#time.year

time1 = pd.date_range(start="1910-01-01", periods=110, freq="1Y")
time1.year


Int64Index([1910, 1911, 1912, 1913, 1914, 1915, 1916, 1917, 1918, 1919,
            ...
            2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019],
           dtype='int64', length=110)

In [20]:
no_data_state = []
no_data_county= []
for i in range(len(county)): #len(county)
  for j in range(len(county[i])): #len(county[i])
    if len(np.atleast_1d(county[i]['corn_yld_yrs'][j])) != 0 or len(np.atleast_1d(county[i]['corn_area_yrs'][j])) !=0:
      corn_yr = np.atleast_1d(county[i]['corn_yld_yrs'][j]) #.astype('int64')
      corn_yld = np.atleast_1d(county[i]['corn_yld'][j]) #.astype('int64')
      corn_area_yr =np.atleast_1d(county[i]['corn_area_yrs'][j]) #.astype('int64')
      corn_area = np.atleast_1d(county[i]['corn_area'][j]) #.astype('int64')
      temp1  = np.zeros([110])
      temp2 = np.zeros([110])
      index1 = []
      index2 = []
      idx1 = 0
      idx2 = 0
      print(i,j)
      for z,k in zip_longest(corn_yr,corn_area_yr):
        if z != None:
          idx1 = np.where(list(time.year) == z)
          index1.append(idx1)
        if k != None:
          idx2= np.where(list(time.year) == k)
          index2.append(idx2)
      #print(np.ravel(index1), np.ravel(index2))
      if len(corn_yr) !=0:
        temp1[np.ravel(index1)] = corn_yld
      if len(corn_area_yr) !=0:
        temp2[np.ravel(index2)] = corn_area
      #print(temp1[np.ravel(index1)], corn_yld,county[i]["mLat"][j])

      # Saving: change from dict to dataframe
      my_dict = dict(state_name=data['st']['name'][i],state_id=data['st']['FIPS'][i],
                    county_name=county[i]['name'][j],county_id=county[i]['ID'][j],
                    lat=county[i]["mLat"][j],lon=county[i]["mLon"][j],
                    year=time1,corn_yld=temp1,corn_area=temp2)
      dataframe_temp = pd.DataFrame.from_dict(my_dict)
      
      if i == 0 and j == 0:
        dataframe = dataframe_temp
      else:
        dataframe = dataframe.append(dataframe_temp,ignore_index=True, sort=False)
    
    else:
      no_data_county.append(j)
      no_data_state.append(i)
      print('no data: ',i,j)




0 0
0 1
0 2
0 3
0 4
0 5
0 6
0 7
0 8
0 9
0 10
0 11
0 12
0 13
0 14
0 15
0 16
0 17
0 18
0 19
0 20
0 21
0 22
0 23
0 24
0 25
0 26
0 27
0 28
0 29
0 30
0 31
0 32
0 33
0 34
0 35
0 36
0 37
0 38
0 39
0 40
0 41
0 42
0 43
0 44
0 45
0 46
0 47
0 48
0 49
0 50
0 51
0 52
0 53
0 54
0 55
0 56
0 57
0 58
0 59
0 60
0 61
0 62
0 63
0 64
0 65
0 66
1 0
1 1
1 2
1 3
1 4
1 5
1 6
1 7
1 8
1 9
1 10
1 11
1 12
1 13
1 14
2 0
2 1
2 2
2 3
2 4
2 5
2 6
2 7
2 8
2 9
2 10
2 11
2 12
2 13
2 14
2 15
2 16
2 17
2 18
2 19
2 20
2 21
2 22
2 23
2 24
2 25
2 26
2 27
2 28
2 29
2 30
2 31
2 32
2 33
2 34
2 35
2 36
2 37
2 38
2 39
2 40
2 41
2 42
2 43
2 44
2 45
2 46
2 47
2 48
2 49
2 50
2 51
2 52
2 53
2 54
2 55
2 56
2 57
2 58
2 59
2 60
2 61
2 62
2 63
2 64
2 65
2 66
2 67
2 68
2 69
2 70
2 71
2 72
2 73
2 74
3 0
no data:  3 1
3 2
3 3
3 4
3 5
3 6
no data:  3 7
3 8
3 9
3 10
3 11
3 12
3 13
3 14
3 15
3 16
3 17
3 18
3 19
no data:  3 20
no data:  3 21
3 22
3 23
3 24
no data:  3 25
3 26
3 27
no data:  3 28
3 29
3 30
no data:  3 31
3 32
3 33
3 34
3 35
3 36


In [9]:
len(no_data_county), len(np.unique(dataframe['lat'])), 277 + 2831 #no_data_county

(277, 2831, 3108)

In [ ]:
pd.to_datetime('1981')

In [22]:
dataframe.loc[(dataframe['year'] > pd.to_datetime('1980')) & (dataframe['year'] < pd.to_datetime('2019'))]

state_name  state_id county_name  ...       year  corn_yld  corn_area
70             AL         1     Autauga  ... 1980-12-31      20.8     6600.0
71             AL         1     Autauga  ... 1981-12-31      17.3     5200.0
72             AL         1     Autauga  ... 1982-12-31      59.1     2200.0
73             AL         1     Autauga  ... 1983-12-31      38.0     2000.0
74             AL         1     Autauga  ... 1984-12-31      41.2     2500.0
...           ...       ...         ...  ...        ...       ...        ...
311404         WY        56      Weston  ... 2014-12-31       0.0        0.0
311405         WY        56      Weston  ... 2015-12-31       0.0        0.0
311406         WY        56      Weston  ... 2016-12-31       0.0        0.0
311407         WY        56      Weston  ... 2017-12-31       0.0        0.0
311408         WY        56      Weston  ... 2018-12-31       0.0        0.0

[110409 rows x 9 columns]

In [21]:
dataframe.to_csv('/content/drive/My Drive/state_county_corn_pd_date.csv')

# 2. PRISM Data with Shape file
### 2-1. Prism Data (2 variables)
<li> source: PRISM () </li>
<li> temporal period: 1981~2016 (36yrs) </li>
<li> spatial scale: county level - 3220 counties (check spatial resolution) </li>
<li> temperature: yearly mean (unit: celius) - shape: (3220,36) </li>
<li> precipitation: yearly accumulated (unit: mm) - shape: (3220,36)</li>
<li> note: compared to crop data, there are more data available. crop data has a total of 3108 counties. </li>
<li> note: spatial coordinates follow cb_2016_us_county_20m/cb_2016_us_county_20m.shp. This is GIS file with polygon and multipolygon. I extracted the central coordinates out of them which seem similar to the representative coordinates of USDA-NASS crop data. </li>
<li> note: missing data as nan </li>

## 2-2. Shapefile (state id, county id, coordinates info to match with PRISM data)
<li> GIS file </li>
<li> The order is the same as the order of PRISM data </li>


In [1]:
!pip install geopandas
import geopandas as gpd

     |████████████████████████████████| 972kB 2.8MB/s 
     |████████████████████████████████| 14.8MB 257kB/s 
     |████████████████████████████████| 10.9MB 40.5MB/s 


##2-1. prism data

In [4]:
mon_temp = scipy.io.loadmat('/content/drive/My Drive/ASCI Lab Data/USCountyDataSummaries/Monthly/tdmean_prism.mat')
mon_precip = scipy.io.loadmat('/content/drive/My Drive/ASCI Lab Data/USCountyDataSummaries/Monthly/ppt_prism.mat')

#data structure
np.shape(mon_temp['tdmean_prism']), np.shape(mon_precip['ppt_prism']), mon_temp['year_list']

((3220, 12, 36),
 (3220, 12, 36),
 array([[1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991,
         1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002,
         2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013,
         2014, 2015, 2016]], dtype=uint16))

In [5]:
time1 = pd.date_range(start="1981-01-01", end='2017-01-01', freq="1Y")
time1

DatetimeIndex(['1981-12-31', '1982-12-31', '1983-12-31', '1984-12-31',
               '1985-12-31', '1986-12-31', '1987-12-31', '1988-12-31',
               '1989-12-31', '1990-12-31', '1991-12-31', '1992-12-31',
               '1993-12-31', '1994-12-31', '1995-12-31', '1996-12-31',
               '1997-12-31', '1998-12-31', '1999-12-31', '2000-12-31',
               '2001-12-31', '2002-12-31', '2003-12-31', '2004-12-31',
               '2005-12-31', '2006-12-31', '2007-12-31', '2008-12-31',
               '2009-12-31', '2010-12-31', '2011-12-31', '2012-12-31',
               '2013-12-31', '2014-12-31', '2015-12-31', '2016-12-31'],
              dtype='datetime64[ns]', freq='A-DEC')

## 2-2. shape data

In [6]:
lonlatinfo = gpd.read_file("/content/drive/My Drive/ASCI Lab Data/USCountyDataSummaries/cb_2016_us_county_20m/cb_2016_us_county_20m.shp")
lonlatinfo.head()

STATEFP COUNTYFP  ...    AWATER                                           geometry
0      39      131  ...   9567612  POLYGON ((-83.35353 39.19758, -82.78589 39.168...
1      46      003  ...  11201379  POLYGON ((-98.80777 43.93522, -98.33151 43.937...
2      55      035  ...  18848512  POLYGON ((-91.65045 44.85595, -90.92225 44.857...
3      72      145  ...  57805868  POLYGON ((-66.44899 18.38721, -66.43895 18.481...
4      48      259  ...   1496797  POLYGON ((-98.92015 30.13829, -98.58790 30.138...

[5 rows x 10 columns]

In [7]:
lonlatinfo.sort_values(by=['STATEFP','COUNTYFP'])

STATEFP  ...                                           geometry
1158      01  ...  POLYGON ((-86.91759 32.66417, -86.71339 32.661...
2846      01  ...  POLYGON ((-88.02632 30.75336, -87.94455 30.827...
1115      01  ...  POLYGON ((-85.73573 31.62449, -85.66623 31.772...
550       01  ...  POLYGON ((-87.42194 33.00338, -87.31854 33.006...
2012      01  ...  POLYGON ((-86.96336 33.85822, -86.92439 33.909...
...      ...  ...                                                ...
3         72  ...  POLYGON ((-66.44899 18.38721, -66.43895 18.481...
1508      72  ...  POLYGON ((-65.57686 18.10322, -65.50592 18.152...
601       72  ...  POLYGON ((-66.52190 18.15195, -66.47055 18.177...
1881      72  ...  POLYGON ((-66.01024 18.07556, -65.93317 18.093...
2352      72  ...  POLYGON ((-66.92652 18.15127, -66.85678 18.138...

[3220 rows x 10 columns]

In [8]:
#from shapely.geometry import Point, Polygon
lonlatinfo['Center_point'] = lonlatinfo['geometry'].centroid
#Extract lat and lon from the centerpoint
lonlatinfo["lat"] = lonlatinfo.Center_point.map(lambda p: p.x)
lonlatinfo["lon"] = lonlatinfo.Center_point.map(lambda p: p.y)


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: UserWarning: Geometry is in a geographic CRS. Results from 'centroid' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

  


In [10]:
test = lonlatinfo.dissolve(by='STATEFP')
test.dissolve(by='COUNTYFP')
test[:10]

geometry  ...        lon
STATEFP                                                     ...           
01       POLYGON ((-85.12544 31.76297, -85.11893 31.732...  ...  31.125679
02       MULTIPOLYGON (((-179.17426 51.27906, -178.9955...  ...  58.744262
04       POLYGON ((-109.04761 32.42638, -109.04830 32.0...  ...  33.214988
05       POLYGON ((-90.94480 34.11666, -90.94632 34.109...  ...  33.212307
06       MULTIPOLYGON (((-118.59403 33.03595, -118.5400...  ...  37.752151
08       POLYGON ((-102.04189 37.64428, -102.04194 37.3...  ...  39.331338
09       POLYGON ((-73.69594 41.11526, -73.48271 41.212...  ...  41.267233
10       POLYGON ((-75.30665 38.94766, -75.30255 38.939...  ...  39.579943
11       POLYGON ((-77.11976 38.93434, -77.04102 38.995...  ...  38.904178
12       MULTIPOLYGON (((-82.01491 24.54307, -81.98391 ...  ...  27.186779

[10 rows x 12 columns]

## 2-3. processing two data

In [12]:
yr_mean_temp = np.mean(mon_temp['tdmean_prism'], axis=1)
yr_mean_temp.shape

yr_accu_pr = np.sum(mon_precip['ppt_prism'],axis=1)
yr_accu_pr.shape 

(3220, 36)

In [13]:
for i in range(3220):
  my_dict = dict(state_id =lonlatinfo['STATEFP'][i], county_name = lonlatinfo['NAME'][i], county_id=lonlatinfo['COUNTYFP'][i],
                 year=time1,yearly_mean_temp = yr_mean_temp[i,:], yearly_accum_precip=yr_accu_pr[i,:])
  dataframe_temp = pd.DataFrame.from_dict(my_dict)
      
  if i == 0:
    dataframe1 = dataframe_temp
  else:
    dataframe1 = dataframe1.append(dataframe_temp,ignore_index=True, sort=False)

In [14]:
dataframe1.sort_values(by =['state_id','county_id'])

state_id county_name  ... yearly_mean_temp yearly_accum_precip
41688       01     Autauga  ...         9.844607         1204.030461
41689       01     Autauga  ...        11.243726         1487.625515
41690       01     Autauga  ...         9.834751         1772.045745
41691       01     Autauga  ...        10.492126         1228.949999
41692       01     Autauga  ...        10.961063         1082.190461
...        ...         ...  ...              ...                 ...
84703       72       Yauco  ...              NaN                 NaN
84704       72       Yauco  ...              NaN                 NaN
84705       72       Yauco  ...              NaN                 NaN
84706       72       Yauco  ...              NaN                 NaN
84707       72       Yauco  ...              NaN                 NaN

[115920 rows x 6 columns]

In [ ]:
dataframe1.to_csv('/content/drive/My Drive/prism_yearly_weather_with_lonlatfromshape.csv')